File: lib/Models/ChatModels.dart
import 'BaseResponse.dart';

// ==================== ChatMessageModel ====================
class ChatMessageModel {
  ChatMessageModel({
    required this.id,
    required this.conversationId,
    required this.senderType,
    this.senderName,
    this.senderId,
    required this.content,
    required this.createdAt,
  });

  final String id;
  final String conversationId;
  final String senderType;
  final String? senderName;
  final String? senderId;
  final String content;
  final DateTime createdAt;

  bool get isAdmin => senderType == 'ADMIN';

  factory ChatMessageModel.fromJson(Map<String, dynamic> json) {
    return ChatMessageModel(
      id: json['id'] as String,
      conversationId: json['conversationId'] as String,
      senderType: json['senderType'] as String,
      senderName: json['senderName'] as String?,
      senderId: json['senderId'] as String?,
      content: json['content'] as String,
      createdAt: DateTime.parse(json['createdAt'] as String),
    );
  }
}

// ==================== ChatConversationModel ====================
class ChatConversationModel {
  ChatConversationModel({
    required this.id,
    required this.userId,
    required this.username,
    this.userEmail,
    this.avatar,
    this.lastMessageAt,
    this.lastMessagePreview,
    this.messages,
  });

  final String id;
  final String userId;
  final String username;
  final String? userEmail;
  final String? avatar;
  final DateTime? lastMessageAt;
  final String? lastMessagePreview;
  final List<ChatMessageModel>? messages;

  factory ChatConversationModel.fromJson(Map<String, dynamic> json) {
    final messages = json['messages'] as List<dynamic>?;
    return ChatConversationModel(
      id: json['id'] as String,
      userId: json['userId'] as String,
      username: json['username'] as String,
      userEmail: json['userEmail'] as String?,
      avatar: json['avatar'] as String?,
      lastMessageAt: json['lastMessageAt'] != null
          ? DateTime.parse(json['lastMessageAt'])
          : null,
      lastMessagePreview: json['lastMessagePreview'] as String?,
      messages: messages
          ?.map((e) => ChatMessageModel.fromJson(e as Map<String, dynamic>))
          .toList(),
    );
  }
}

// ==================== ChatConversationResponseWrapper ====================
class ChatConversationResponseWrapper extends BaseResponse {
  ChatConversationModel? data;

  ChatConversationResponseWrapper({
    required int statusCode,
    required String message,
    this.data,
  }) : super(statusCode: statusCode, message: message);

  factory ChatConversationResponseWrapper.fromJson(Map<String, dynamic> json) {
    final payload = json['data'];
    return ChatConversationResponseWrapper(
      statusCode: (json['statusCode'] as num).toInt(),
      message: json['message'] as String,
      data: payload is Map<String, dynamic>
          ? ChatConversationModel.fromJson(payload)
          : null,
    );
  }
}

// ==================== ChatMessageResponseWrapper ====================
class ChatMessageResponseWrapper extends BaseResponse {
  ChatMessageModel? data;

  ChatMessageResponseWrapper({
    required int statusCode,
    required String message,
    this.data,
  }) : super(statusCode: statusCode, message: message);

  factory ChatMessageResponseWrapper.fromJson(Map<String, dynamic> json) {
    final payload = json['data'];
    return ChatMessageResponseWrapper(
      statusCode: (json['statusCode'] as num).toInt(),
      message: json['message'] as String,
      data:
          payload is Map<String, dynamic> ? ChatMessageModel.fromJson(payload) : null,
    );
  }
}

File: mobile_restaurant/models/ChatModels.ts
import { parseBaseResponse, type BaseResponse } from './AuthModels';

// ==================== ChatMessageModel ====================
export interface ChatMessageModel {
  id: string;
  conversationId: string;
  senderType: string;
  senderName?: string;
  senderEmail?: string;
  senderId?: string;
  content: string;
  createdAt: string;
  isAdmin: boolean;
}

export function parseChatMessageModel(json: any): ChatMessageModel {
  return {
    id: json.id,
    conversationId: json.conversationId,
    senderType: json.senderType,
    senderName: json.senderName ?? undefined,
    senderEmail: json.senderEmail ?? undefined,
    senderId: json.senderId ?? undefined,
    content: json.content,
    createdAt: json.createdAt,
    isAdmin: json.senderType === 'ADMIN',
  };
}

// ==================== ChatConversationModel ====================
export interface ChatConversationModel {
  id: string;
  userId: string;
  username: string;
  userEmail?: string;
  avatar?: string;
  lastMessageAt?: string;
  lastMessagePreview?: string;
  messages?: ChatMessageModel[];
  status?: string;
}

export function parseChatConversationModel(json: any): ChatConversationModel {
  return {
    id: json.id,
    userId: json.userId,
    username: json.username,
    userEmail: json.userEmail ?? undefined,
    avatar: json.avatar ?? undefined,
    lastMessageAt: json.lastMessageAt ?? undefined,
    lastMessagePreview: json.lastMessagePreview ?? undefined,
    messages: json.messages?.map(parseChatMessageModel) ?? undefined,
    status: json.status ?? undefined,
  };
}

// ==================== ChatConversationResponseWrapper ====================
export interface ChatConversationResponseWrapper extends BaseResponse {
  data?: ChatConversationModel;
}

export function parseChatConversationResponseWrapper(json: any): ChatConversationResponseWrapper {
  const base = parseBaseResponse(json);
  return {
    ...base,
    data: json.data && typeof json.data === 'object' ? parseChatConversationModel(json.data) : undefined,
  };
}

// ==================== ChatMessageResponseWrapper ====================
export interface ChatMessageResponseWrapper extends BaseResponse {
  data?: ChatMessageModel;
}

export function parseChatMessageResponseWrapper(json: any): ChatMessageResponseWrapper {
  const base = parseBaseResponse(json);
  return {
    ...base,
    data: json.data && typeof json.data === 'object' ? parseChatMessageModel(json.data) : undefined,
  };
}
